File: packages/autonomo-python/autonomo/actions.py
"""
Custom Actions - Fast-path operations for complex interactions

Some operations (like OTP entry) require multiple steps that are
slow and flaky when done individually. Custom actions provide
atomic operations that handle these cases.
"""

from dataclasses import dataclass, field
from typing import Callable, Optional, Dict, Any, List


@dataclass
class ActionResult:
    """Result of a custom action"""
    success: bool
    message: Optional[str] = None
    error: Optional[str] = None
    data: Optional[Any] = None

    def to_dict(self) -> Dict[str, Any]:
        result = {"success": self.success}
        if self.message is not None:
            result["message"] = self.message
        if self.error is not None:
            result["error"] = self.error
        if self.data is not None:
            result["data"] = self.data
        return result


@dataclass
class CustomActionMeta:
    """Metadata for a custom action - helps AI understand what it does"""
    description: Optional[str] = None
    args: Optional[Dict[str, str]] = None
    example: Optional[Dict[str, str]] = None


@dataclass
class CustomActionInfo:
    """Rich custom action info returned in state"""
    name: str
    description: Optional[str] = None
    args: Optional[Dict[str, str]] = None
    example: Optional[Dict[str, str]] = None

    def to_dict(self) -> Dict[str, Any]:
        result: Dict[str, Any] = {"name": self.name}
        if self.description is not None:
            result["description"] = self.description
        if self.args is not None:
            result["args"] = self.args
        if self.example is not None:
            result["example"] = self.example
        return result


@dataclass
class RegisteredAction:
    """Internal representation of a registered action"""
    handler: Callable[[Optional[str]], ActionResult]
    meta: Optional[CustomActionMeta] = None


CustomActionHandler = Callable[[Optional[str]], ActionResult]


class CustomActionsRegistry:
    """Singleton registry for custom actions"""
    
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._actions: Dict[str, RegisteredAction] = {}
            cls._instance._listeners: List[Callable[[], None]] = []
        return cls._instance
    
    def register(
        self, 
        name: str, 
        handler: CustomActionHandler,
        meta: Optional[CustomActionMeta] = None
    ) -> Callable[[], None]:
        """Register a custom action. Returns unregister function."""
        self._actions[name] = RegisteredAction(handler=handler, meta=meta)
        self._notify_change()
        return lambda: self.unregister(name)
    
    def unregister(self, name: str) -> None:
        """Unregister a custom action"""
        if name in self._actions:
            del self._actions[name]
            self._notify_change()
    
    def execute(self, name: str, value: Optional[str] = None) -> ActionResult:
        """Execute a custom action"""
        action = self._actions.get(name)
        if action is None:
            available = ", ".join(self.list()) if self._actions else "none"
            return ActionResult(
                success=False,
                error=f"Unknown custom action: {name}. Available: {available}",
            )
        try:
            return action.handler(value)
        except Exception as err:
            return ActionResult(
                success=False,
                error=str(err),
            )
    
    def has(self, name: str) -> bool:
        """Check if action exists"""
        return name in self._actions
    
    def list(self) -> List[str]:
        """List all action names"""
        return list(self._actions.keys())
    
    def get_all(self) -> List[CustomActionInfo]:
        """Get rich info about all actions (for AI discoverability)"""
        result = []
        for name, action in self._actions.items():
            info = CustomActionInfo(
                name=name,
                description=action.meta.description if action.meta else None,
                args=action.meta.args if action.meta else None,
                example=action.meta.example if action.meta else None,
            )
            result.append(info)
        return result
    
    def on_change(self, listener: Callable[[], None]) -> Callable[[], None]:
        """Subscribe to changes"""
        self._listeners.append(listener)
        return lambda: self._listeners.remove(listener)
    
    def _notify_change(self) -> None:
        for listener in self._listeners:
            listener()


# Singleton instance
custom_actions = CustomActionsRegistry()


def register_custom_action(
    name: str,
    handler: CustomActionHandler,
    meta: Optional[CustomActionMeta] = None,
) -> Callable[[], None]:
    """Register a custom action"""
    return custom_actions.register(name, handler, meta)

File: packages/autonomo-core/src/actions.ts
/**
 * Custom Actions - Fast-path operations for complex interactions
 * 
 * Some operations (like OTP entry) require multiple steps that are
 * slow and flaky when done individually. Custom actions provide
 * atomic operations that handle these cases.
 */

export type CustomActionHandler = (
  value?: string
) => Promise<ActionResult> | ActionResult;

export interface ActionResult {
  success: boolean;
  message?: string;
  error?: string;
  data?: unknown;
}

/**
 * Metadata for a custom action - helps AI understand what it does
 */
export interface CustomActionMeta {
  /** Human-readable description of what the action does */
  description?: string;
  /** Argument schema: { argName: 'type description' } */
  args?: Record<string, string>;
  /** Example usage */
  example?: { value?: string };
}

/**
 * Rich custom action info returned in state
 */
export interface CustomActionInfo {
  name: string;
  description?: string;
  args?: Record<string, string>;
  example?: { value?: string };
}

type ActionsChangeListener = () => void;

interface RegisteredAction {
  handler: CustomActionHandler;
  meta?: CustomActionMeta;
}

class CustomActionsRegistry {
  private actions = new Map<string, RegisteredAction>();
  private listeners = new Set<ActionsChangeListener>();

  /**
   * Register a custom action
   * @param name - Action name (e.g., 'fillOtp')
   * @param handler - Function to execute the action
   * @param meta - Optional metadata describing the action for AI
   * @returns Unregister function
   */
  register(name: string, handler: CustomActionHandler, meta?: CustomActionMeta): () => void {
    this.actions.set(name, { handler, meta });
    this.notifyChange();
    return () => this.unregister(name);
  }

  /**
   * Unregister a custom action
   */
  unregister(name: string): void {
    this.actions.delete(name);
    this.notifyChange();
  }

  /**
   * Execute a custom action
   */
  async execute(name: string, value?: string): Promise<ActionResult> {
    const action = this.actions.get(name);
    if (!action) {
      return {
        success: false,
        error: `Unknown custom action: ${name}. Available: ${this.list().join(', ') || 'none'}`,
      };
    }
    try {
      return await action.handler(value);
    } catch (err) {
      return {
        success: false,
        error: err instanceof Error ? err.message : String(err),
      };
    }
  }

  /**
   * Check if action exists
   */
  has(name: string): boolean {
    return this.actions.has(name);
  }

  /**
   * List all action names
   */
  list(): string[] {
    return Array.from(this.actions.keys());
  }

  /**
   * Get rich info about all actions (for AI discoverability)
   */
  getAll(): CustomActionInfo[] {
    return Array.from(this.actions.entries()).map(([name, action]) => ({
      name,
      description: action.meta?.description,
      args: action.meta?.args,
      example: action.meta?.example,
    }));
  }

  /**
   * Subscribe to changes
   */
  onChange(listener: ActionsChangeListener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notifyChange(): void {
    this.listeners.forEach(listener => listener());
  }
}

// Singleton instance
export const customActions = new CustomActionsRegistry();

// Convenience function
export function registerCustomAction(
  name: string,
  handler: CustomActionHandler,
  meta?: CustomActionMeta
): () => void {
  return customActions.register(name, handler, meta);
}
